Update documentation for sections (#187)
fofr/govuk-design-history: app/posts/2020-01-01-divide-a-design-history-into-different-sections.md

date: 2020-01-01
---

> If your service has different interfaces (for example a public facing service as well as an admin or support interface), you may want to divide your history up into different sections focused on each part of the service.

Reviewer: ```suggestion
If your service has different parts, for example a public facing service and an admin interface, you may want to divide your history into sections for each part.
```

Author: Nice, thanks!

## Set up a folder for each section

> You can group posts together into sections using folders

Reviewer: ```suggestion
You can group posts together into sections using folders:
```

> 1. Create a subfolder for each section within the app/posts folder. Use dashes instead of spaces in the the folder names.

Reviewer: ```suggestion
1. Create a folder for each section in `app/posts`. Use dashes instead of spaces in the the folder names.
```

> 2. Within each subfolder, create a JSON file, with the same name as the folder but ending in `.json`. For example, for the folder `app/posts/support-interface`, add the file `app/posts/support-interface/support-interface.json`.

Reviewer: ```suggestion
2. Create a `.json` file in each folder. Use the same name as the folder. For the folder `app/posts/support-interface`, the JSON file would be `app/posts/support-interface/support-interface.json`.
```

Next, create a page that lists these related posts. You can do that by creating an index page.

> 1. Within the `posts` folder, create a markdown file for each section. For example, for the folder `app/posts/support-interface/`, add the file `app/posts/support-interface.md`.

Reviewer: ```suggestion
1. In the `posts` folder, create a markdown `.md` file for each section. For the folder `app/posts/support-interface/`, add the file `app/posts/support-interface.md`.
```

> 1. Within the `posts` folder, create a markdown file for each section. For example, for the folder `app/posts/support-interface/`, add the file `app/posts/support-interface.md`.

Reviewer: Why the `posts` folder and not the subfolder?

For the TRA design history they were in the subfolder?
https://github.com/DFE-Digital/tra-digital-design-history/blob/main/app/posts/get-an-identity/get-an-identity.md

Or does this simplify the required markdown?

Author: So when I tried following the instructions, creating a file in the folder with the same name as the folder didn't work - the resulting URL used the name twice (`get-an-identity/get-an-identity`) and didn't seem to show up on the homepage. Either I got something wrong, or Eleventy (or the govuk plugin?) has changed since the instructions.

I think intuitively the index page to the top level `posts/` folder makes more sense to me, as it's easier to spot, and follows the URL path a bit more closely.  The other traditional approach would be a `index.md` file within the subfolder, but that doesn’t seem to be the conventional Eleventy setup?

Reviewer 2: `index.md` files in folders containing directory data cause all sorts of issues, as you typically need to override the tags and other data for errors not to occur. This is a major pain with Eleventy to be honest (and this is why @fofr would be using `override:tags:` in his collection).

Reviewer 2: In Paul’s example above, the `index.md` page also has a `permalink` value:

`permalink: get-an-identity/{% if pagination.pageNumber > 0 %}page/{{ pagination.pageNumber + 1 }}{% endif %}/`

This would avoid the name duplication. Does the documentation include this?

Author: Ah! So I just tried this locally, and even when adding all these extra bits to the front matter:

```yaml
override:tags: []
eleventyComputed:
  eleventyNavigation:
    key: "{{ title }}"
    excerpt: "{{ description }}"
    parent: home
permalink: "service1/{% if pagination.pageNumber > 0 %}page/{{ pagination.pageNumber + 1 }}{% endif %}/"
```

...something didn't work, I get this error when starting the app:

```bash
1. Having trouble writing template: "public/service1/index.html" (via EleventyTemplateError)
2. (./app/_layouts/collection.njk)
  Error: Node does not exist: Service 1 (via Template render error)
```

Probably I'm missing something else...

Do we think it’d be simpler to switch to suggesting adding the collection pages in the `posts/` folder, or should we continue to keep it in the subfolder, and document all the overrides?

Reviewer: If the collection page is in `posts/`, does it just work, without any overrides?

Author: > If the collection page is in posts/, does it just work, without any overrides?

Yeah, I don't think you need any overrides then, as the json file only applies tags and layouts etc to all the files in the subfolder, so it's a bit simpler...

Reviewer 2: > Do we think it’d be simpler to switch to suggesting adding the collection pages in the posts/ folder

Yes.

> 1. Within the `posts` folder, create a markdown file for each section. For example, for the folder `app/posts/support-interface/``, add the file `app/posts/support-interface.md`.

Reviewer 2: Markdown typo

```suggestion
1. Within the `posts` folder, create a markdown file for each section. For example, for the folder `app/posts/support-interface/`, add the file `app/posts/support-interface.md`.
```

     reverse: true
     size: 50
   permalink: "support-interface/{% if pagination.pageNumber > 0 %}page/{{ pagination.pageNumber + 1 }}{% endif %}/"
   ---
   ```

   {% endraw %}

>    The label appearing after `date: collections.` needs to match the tag you added to the JSON file for the section. This tells the index page which set of posts to list.

Reviewer 2: Bit more pedantic possibly, but I also got confused by `date: collections.`. Also, does a collection with a dash in the name not need to use the bracketed notation, i.e. `collections['support-interface']. Not sure.

```suggestion
   The value for `pagination.data` should use the collection created by the tag you added to the JSON file for the section. This tells the index page which posts to list. For example, if your tag is `support-interface`, the value for `pagination.data` would be `collections.support-interface`.
```

Author: Thanks - your version is much better. I wasn't sure how to explain this easily. (Didn't help that I typoed `data` as `date` 🤦 ).

I just checked and hyphens are fine to use within the frontmatter - `data: collections.support-interface` works, and even spaces seem to work somehow! (I guess the front matter syntax is based on YAML?)
